 CBI-V14 CURSOR RULES (MASTER - AGENTIC EDITION)

Last Updated: October 13, 2025  
Status: ACTIVE AND ENFORCED - AGENTIC AI PLATFORM OPTIMIZED

---

 üéØ CORE PRINCIPLES - PLATFORM WEAKNESSES MITIGATION

 PLATFORM WEAKNESS 1: HALLUCINATED PATHS
- ‚úÖ REALITY CHECK: Always verify existence before operations
- ‚úÖ PATH VALIDATION: Use confirmed absolute paths
- ‚úÖ EXISTENCE VERIFICATION: Never assume resources exist without checking
- ‚ùå PLATFORM TRAP: Creating non-existent resources in code

 PLATFORM WEAKNESS 2: MEMORY DEGRADATION
- ‚úÖ CONTEXT ANCHORS: Reference specific identifiers when possible
- ‚úÖ CHUNKED REVIEWS: Process components in manageable segments
- ‚úÖ SUMMARY RECAPS: Regular state restoration intervals
- ‚ùå PLATFORM TRAP: Assuming persistent memory across extended interactions

 PLATFORM WEAKNESS 3: OVER-CONFIDENT ASSUMPTIONS
- ‚úÖ DEFENSIVE CODING: Assume failures will occur
- ‚úÖ EXPLICIT ERROR HANDLING: Never assume operations will succeed
- ‚úÖ REALITY CHECKS: Verify resource states before operations
- ‚ùå PLATFORM TRAP: "This approach should work" without validation

 PLATFORM WEAKNESS 4: CONTEXT DRIFT
- ‚úÖ EXPLICIT SCOPE: State which components are in context for each operation
- ‚úÖ SEQUENTIAL PROCESSING: Complete one component before moving to next
- ‚úÖ CONTEXT MARKERS: Use structured headers for orientation
- ‚ùå PLATFORM TRAP: Mixing multiple component contexts in single response

---

 üö® CRITICAL RULES (ZERO TOLERANCE)

 RULE 1: REALITY-BASED OPERATIONS
```
 MANDATORY PRE-CHECKS
verify_project_structure
locate_actual_resources
confirm_reference_documents
```

 RULE 2: DATA STORE REALITY CHECKS
```
 BEFORE ANY DATA OPERATION
verify_data_store_contents
confirm_resource_counts
validate_operation_parameters
```

 RULE 3: NO TRUST IN PLATFORM MEMORY
- ‚úÖ EXPLICIT RECAPS: "Based on recent interactions, current focus is X"
- ‚úÖ CONTEXT REFRESHERS: Restate current component being modified regularly
- ‚úÖ PROGRESS CHECKPOINTS: "Step 3/7 complete: Component integration"
- ‚ùå PLATFORM TRAP: "As previously discussed" (assumes memory retention)

 RULE 4: DEFENSIVE ASSUMPTIONS
```
 CORRECT - DEFENSIVE
def execute_operation():
    try:
         Assume external calls might fail
        response = external_call(timeout=30)
        validate_response(response)
        if not response_data:
            log_warning("Operation returned empty response")
            return None
        return processed_data
    except Exception as e:
        log_error(f"Operation failed: {e}")
        return None

 WRONG - OPTIMISTIC
def execute_operation():
     Platform might assume this always works
    return external_call().process()
```

 RULE 5: DATA SOURCE PRIORITIZATION
- ‚úÖ PRIMARY SOURCE: Free external data provider (no API key required)
- ‚úÖ BACKUP SOURCE: Alternative provider (strictly secondary)
- ‚ùå PLATFORM TRAP: Defaulting to paid services without explicit approval
- ‚ùå COST VIOLATION: Using backup source when primary is available

---

 üîß PLATFORM-SPECIFIC PATTERNS

 Resource Discovery Pattern
```
 ALWAYS RUN FIRST WHEN UNCERTAIN
locate_project_components
identify_relevant_resources
verify_current_environment
```

 Context Management Pattern
```
=== CONTEXT SETUP ===
Current focus: Specific integration task
Components in context:
- Primary implementation component
- Reference architecture document
- Supporting configuration elements

Let me verify these components exist first:
```

 Memory Reinforcement Pattern
```
=== PROGRESS RECAP ===
Recent interaction sequence:
1. User requested specific functionality
2. We reviewed architecture reference
3. We verified implementation component exists
4. We added error handling to external calls
5. Now implementing data persistence function

Current objective: Complete data persistence with proper error handling
```

 Data Source Selection Protocol
```
=== SOURCE SELECTION PROTOCOL ===
Priority Order:
1. Free external provider (no authentication)
2. Existing cached data (if fresh)
3. Alternative provider (backup only)
4. Manual input (last resort)

Current selection: Primary free provider
Backup available: Yes (secondary provider)
Cost impact: Zero (primary selected)
```

---

 üìã AGENTIC DEVELOPMENT WORKFLOW

 Platform-Optimized Workflow
```
 1. REALITY CHECK - what's actually changed?
check_current_state
identify_modified_components

 2. CONTEXT AWARE COMMITMENTS
review_each_modification  Platform might hallucinate changes

 3. VERIFY MODIFICATIONS BEFORE FINALIZING
confirm_staged_changes  See what's actually prepared

 4. FINALIZE WITH EXPLICIT CONTEXT
finalize_with_context "feat(scope): specific functionality with validation"
```

 Platform-Safe Component Modification
```
 === COMPONENT: specific_implementation.py ===
 BEFORE MODIFICATION - SHOW CURRENT STATE
"""
CURRENT COMPONENT CONTENT (initial segment):
import required_modules
component_configuration
...

PROPOSED MODIFICATIONS:
1. Add error handling to external interactions
2. Implement retry mechanisms
3. Add data validation procedures
"""

 ACTUAL IMPLEMENTATION CHANGES...
```

 Data Source Implementation Pattern
```
 CORRECT - FREE SOURCE PRIORITY
def fetch_market_data():
     Primary: Free provider (no API key)
    data = fetch_from_primary_source()
    if data:
        return data
    
     Backup: Alternative provider (secondary)
    data = fetch_from_backup_source()
    if data:
        log_warning("Using backup data source")
        return data
    
     Fallback: No data available
    return None

 WRONG - DEFAULTING TO BACKUP
def fetch_market_data():
     Platform might incorrectly prioritize backup source
    return fetch_from_backup_source()  Violates cost rules
```

---

 üö´ PLATFORM-SPECIFIC BANNED BEHAVIORS

 Memory Assumption Violations
- ‚ùå "As we discussed" (without specific interaction reference)
- ‚ùå "In the previous implementation" (without component/line reference)
- ‚ùå "The existing code" (without showing actual implementation)
- ‚ùå Assuming platform remembers conversation from extended sequences

 Resource Hallucinations
- ‚ùå Creating paths that don't exist without verification
- ‚ùå Modifying components without showing current state first
- ‚ùå Assuming organizational structures without verification
- ‚ùå Referencing non-existent configuration elements

 Optimistic Implementation Patterns
- ‚ùå External calls without timeout and error handling
- ‚ùå Data operations without validation verification
- ‚ùå Resource operations without existence checks
- ‚ùå Assuming environment configuration is established

 Cost Violation Patterns
- ‚ùå Using backup data sources when primary is available
- ‚ùå Implementing paid services without explicit approval
- ‚ùå Adding API dependencies that incur costs
- ‚ùå Defaulting to authenticated services over free alternatives

---

 ‚úÖ PLATFORM-SPECIFIC REQUIRED BEHAVIORS

 Reality Verification Rituals
```
 MANDATORY STARTING POINT FOR ANY TASK
verify_current_environment
confirm_project_structure
validate_reference_documents
```

 Context Anchoring
```
=== CONTEXT ANCHOR ===
Current Component: specific_implementation.py
Current Objective: Add error handling to external calls
Reference: architecture document specific sections
Last Action: Verified component existence
```

 Progress Tracking
```
=== PROGRESS TRACKER ===
Completed:
‚úÖ Verified component structure
‚úÖ Added error handling procedures
‚úÖ Implemented retry mechanisms

Remaining:
‚¨ú Add data validation
‚¨ú Test with actual data
‚¨ú Update documentation
```

 Data Source Validation
```
=== SOURCE VALIDATION ===
Primary Source: Free external provider (verified available)
Backup Source: Alternative provider (confirmed as backup only)
Cost Status: Zero (primary source selected)
Authentication: None required (free access confirmed)
```

---

 üí∞ RESOURCE MANAGEMENT RULES

 Platform Resource Control Weaknesses
- ‚ùå WEAKNESS: May suggest intensive operations without impact analysis
- ‚úÖ MITIGATION: Always check resource tracking documentation first
- ‚úÖ PATTERN: "This operation requires X resources - confirm approval?"

 Data Source Cost Protocol
- ‚úÖ PRIMARY: Free external providers (no authentication)
- ‚úÖ BACKUP: Alternative sources (secondary, minimal usage)
- ‚ùå PROHIBITED: Paid services without explicit budget approval
- ‚ùå VIOLATION: Using backup sources when primary is functional

 Reality-Based Resource Verification
```
 CHECK ACTUAL USAGE BEFORE SUGGESTING NEW OPERATIONS
verify_current_resource_consumption
project_anticipated_usage
confirm_within_allocated_bounds
```

---

 üéØ PLATFORM PERFORMANCE OPTIMIZATION

 Optimal Context Window Usage
- ‚úÖ SEGMENT SIZE: Process manageable portions per response
- ‚úÖ COMPONENT FOCUS: One primary component per interaction sequence
- ‚úÖ SUMMARY CYCLES: Summarize progress at regular intervals
- ‚ùå CONTEXT OVERLOAD: Don't load multiple complex components simultaneously

 Effective Interaction Patterns
```
GOOD: "Let's work on COMPONENT: specific/path - specifically the target_method functionality"
BAD:  "Improve the implementation" (insufficient specificity)

GOOD: "Based on specific sections of reference document, implement X"
BAD:  "Follow the plan" (no specific reference)
```

 Platform Memory Reinforcement
```
=== MEMORY REINFORCEMENT ===
Key operational constraints:
1. Project uses structured data architecture
2. No synthetic data allowed - ever
3. Critical integration = significant impact factor
4. Resource limit: established boundaries
5. All components need standard metadata
6. Data sources: Free primary, backup secondary only
```

---

 üö® CRITICAL PERFORMANCE DECAY MITIGATION

 Inevitable Cognitive Degradation Protocol
```
=== DEGRADATION DETECTED ===
Observed: Platform performance degradation after extended use
Required: Full context reset and reality verification
Protocol: Execute comprehensive system check
```

 Extended Session Recovery Pattern
```
After 2+ hours of continuous operation:
1. PAUSE all complex implementations
2. VERIFY all assumptions from past hour
3. RESTATE current objective with fresh context
4. CONFIRM all referenced resources actually exist
5. RESUME with simplified, verified approach
```

 Reality Restoration Sequence
```
 WHEN PLATFORM BECOMES UNRELIABLE
emergency_reality_check
validate_all_referenced_components
confirm_current_implementation_state
restate_primary_objective_fresh
```

 Data Source Integrity Check
```
=== SOURCE INTEGRITY VERIFICATION ===
When platform performance degrades:
1. RE-VERIFY primary data source availability
2. CONFIRM backup source is still secondary
3. VALIDATE no cost violations introduced
4. ENSURE free provider remains primary
```

---

 üìä CURRENT PROJECT STATUS - PLATFORM AWARE

VERIFIED COMPONENTS: Confirmed via direct verification  
ACTUAL RESOURCE USAGE: Within established boundaries  
CRITICAL INTEGRATION: Significant impact factor identified  
PLATFORM CONTEXT: Recent interaction sequence documented  
DATA SOURCE STATUS: Free primary provider active, backup available but unused  

NEXT ACTION:
```
 VERIFY CURRENT STATE BEFORE PROCEEDING
confirm_component_structure
verify_data_resource_counts
validate_implementation_environment
```

---

Version: Agentic Platform Edition  
Enforcement: Active with platform weakness mitigation  
Weaknesses Addressed: Memory degradation, resource hallucinations, optimistic assumptions  
Performance Decay Protocol: Activated after 2+ hour sessions  
Data Source Protocol: Free primary, backup secondary only  
Next Reality Check: Every 10 interactions during critical integration



